akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala

```diff
@@ -659,6 +861,8 @@ final class ClusterReceptionist(pubSubMediator: ActorRef, settings: ClusterRecep
     case Heartbeat ⇒
       if (verboseHeartbeat) log.debug("Heartbeat from client [{}]", sender().path)
       sender() ! HeartbeatRsp
+      updateClientInteractions(sender())
+      publishClientsUp()
```

Comment:
Perhaps mimimize the amount of work that is done for each heartbeat for the normal case. E.g. 
```scala
if (!clientInteractions.contains(sender())) {
  updateClientInteractions(sender())
  publishClientsUp()
}
```